extensions/amp-viewer-integration/0.1/touch_handler.js

```diff
@@ -0,0 +1,184 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+import {listen} from '../../../src/event-helper';
+
+
+/**
+ * The list of touch event properites to copy.
+ * @const {!Array<string>}
+ */
+const EVENT_PROPERTIES = [
+  'altKey', 'charCode', 'ctrlKey', 'detail', 'eventPhase', 'keyCode',
+  'layerX', 'layerY', 'metaKey', 'pageX', 'pageY', 'returnValue',
+  'shiftKey', 'timeStamp', 'type', 'which',
+];
+
+/**
+ * The list of touch properties to copy.
+ * @const {!Array<string>}
+ */
+const TOUCH_PROPERTIES = [
+  'clientX', 'clientY', 'force', 'identifier', 'pageX', 'pageY', 'radiusX',
+  'radiusY', 'screenX', 'screenY',
+];
+
+/**
+ * @const {string} Request name to enable/disable scrolling.
+ */
+const SCROLL_LOCK = 'scrollLock';
+
+/**
+ * @fileoverview Forward touch events from the AMP doc to the viewer.
+ */
+export class TouchHandler {
+
+  /**
+   * @param {!Window} win
+   * @param {!./messaging.Messaging} messaging
+   */
+  constructor(win, messaging) {
+    /** @const {!Window} */
+    this.win = win;
+    /** @private {!./messaging.Messaging} */
+    this.messaging_ = messaging;
+    /**
+     * Do not forward touch events when false.
+     * @private {boolean}
+     */
+    this.tracking_ = false;
+
+    /**
+     * When true, prevent default to prevent scrolling.
+     * @private {boolean}
+     */
+    this.scrollLocked_ = false;
+
+
+    messaging.registerHandler(SCROLL_LOCK, this.scrollLockHandler_.bind(this));
+
+    this.listenForTouchEvents();
+  }
+
+  listenForTouchEvents() {
+    const handleEvent = this.handleEvent_.bind(this);
+
+    listen(this.win, 'touchstart', handleEvent);
+    listen(this.win, 'touchend', handleEvent);
+    listen(this.win, 'touchmove', handleEvent);
+  }
+
+  /**
+   * @param {!Event} e
+   * @private
+   */
+  handleEvent_(e) {
+    switch (e.type) {
+      case 'touchstart':
+        this.tracking_ = true;
```

Comment:
The Google script does do some detection, which is what it uses `tracking_` for. This was done primarily to support horizontal scrolling within an iframe. This is no longer possible so I think this should work without the detection. It would just be an optimization to add it.